File: Core/Models/Athlete.swift
import Foundation

public enum Gender: String, Codable, CaseIterable, Sendable, Hashable {
    case male, female
}

public enum AgeGroup: String, Codable, CaseIterable, Sendable, Hashable {
    case cubs, kids, cadets, juniors, seniors, masters

    public var labelKey: String { "age.\(rawValue)" }

    public static func from(age: Int) -> AgeGroup {
        switch age {
        case ..<10: .cubs
        case 10...11: .kids
        case 12...14: .cadets
        case 15...17: .juniors
        case 18...39: .seniors
        default: .masters
        }
    }
}

public enum AthleteStatus: String, Codable, CaseIterable, Sendable, Hashable {
    case competitionTeam, readyToGrade, watch, rest, active

    public var labelKey: String { "status.\(rawValue)" }
}

public struct Athlete: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var memberNumber: Int
    public var fullName: String
    public var fullNameAr: String
    public var dateOfBirth: Date
    public var gender: Gender
    public var nationality: String
    public var emiratesID: String?
    public var branchID: EntityID
    public var primaryCoachID: EntityID?
    public var joinedAt: Date
    public var currentBelt: Belt
    public var beltHistory: [Belt]
    public var weightKg: Double
    public var status: AthleteStatus
    public var avatarSeed: String
    public var avatarURL: String?

    // === Identity ===
    public var passportNumber: String?
    public var bloodType: BloodType?
    public var federationLicenceNumber: String?
    /// World Taekwondo Federation athlete ID (GAL system). Distinct from the
    /// UAE federation licence number above.
    public var worldTaekwondoID: String?

    // === Family / consent ===
    public var parentUserIDs: [EntityID]
    public var emergencyContacts: [EmergencyContact]
    public var school: String?
    public var imageRightsConsent: Bool
    public var imageRightsConsentDate: Date?
    public var travelPermission: Bool
    public var travelPermissionDate: Date?

    // === Medical ===
    public var heightCm: Double?
    public var weightHistory: [WeightEntry]
    public var allergies: [String]
    public var medicalConditions: [String]
    public var medications: [String]
    public var fitToTrain: Bool
    public var injuries: [InjuryEntry]

    // === Technical ===
    public var weightClass: WeightCategory?
    public var dominantLeg: DominantLeg?
    public var dominantStance: Stance?
    public var specialty: Specialty?
    public var yearsTraining: Int?
    public var poomsaeSyllabus: String?
    public var kyorugiTier: KyorugiTier?
    public var trainingGroupID: EntityID?

    /// Repertoire — forms the athlete can perform. Mastery scores live on
    /// `PoomsaeAssessment`; this set tracks which forms are in the repertoire
    /// independent of how well they're performed.
    public var poomsaeKnown: Set<PoomsaeForm>

    // === Grading (Pillar 9) ===
    /// 1...5 — coach's subjective readiness rating for the next grading.
    public var gradingReadiness: Int?
    /// Planned date for the athlete's next belt-test.
    public var nextGradingTargetDate: Date?

    // === Profile dossier (Pillar 10) ===
    /// Threaded coach notes. Newest first by convention; the view layer sorts
    /// defensively anyway. Empty in the standard demo seed; populated for the
    /// "showcase" athletes in `SeedData.build()`.
    public var coachNotes: [CoachNote]
    /// Identity / consent / medical documents. Status is computed from the
    /// expiry date at render-time via `derivedStatus(asOf:)`.
    public var documents: [AthleteDocument]
    /// Latest ranking snapshot across club / UAE / WT / Olympic systems.
    /// Nil for non-competitive athletes — the Overview tab hides the
    /// Current Rankings card when this is nil.
    public var ranking: AthleteRanking?

    public init(
        id: EntityID = UUID(),
        memberNumber: Int,
        fullName: String,
        fullNameAr: String,
        dateOfBirth: Date,
        gender: Gender = .male,
        nationality: String = "AE",
        emiratesID: String? = nil,
        branchID: EntityID,
        primaryCoachID: EntityID? = nil,
        joinedAt: Date,
        currentBelt: Belt,
        beltHistory: [Belt] = [],
        weightKg: Double,
        status: AthleteStatus,
        avatarSeed: String,
        avatarURL: String? = nil,
        passportNumber: String? = nil,
        bloodType: BloodType? = nil,
        federationLicenceNumber: String? = nil,
        worldTaekwondoID: String? = nil,
        parentUserIDs: [EntityID] = [],
        emergencyContacts: [EmergencyContact] = [],
        school: String? = nil,
        imageRightsConsent: Bool = false,
        imageRightsConsentDate: Date? = nil,
        travelPermission: Bool = false,
        travelPermissionDate: Date? = nil,
        heightCm: Double? = nil,
        weightHistory: [WeightEntry] = [],
        allergies: [String] = [],
        medicalConditions: [String] = [],
        medications: [String] = [],
        fitToTrain: Bool = true,
        injuries: [InjuryEntry] = [],
        weightClass: WeightCategory? = nil,
        dominantLeg: DominantLeg? = nil,
        dominantStance: Stance? = nil,
        specialty: Specialty? = nil,
        yearsTraining: Int? = nil,
        poomsaeSyllabus: String? = nil,
        kyorugiTier: KyorugiTier? = nil,
        trainingGroupID: EntityID? = nil,
        poomsaeKnown: Set<PoomsaeForm> = [],
        gradingReadiness: Int? = nil,
        nextGradingTargetDate: Date? = nil,
        coachNotes: [CoachNote] = [],
        documents: [AthleteDocument] = [],
        ranking: AthleteRanking? = nil
    ) {
        self.id = id
        self.memberNumber = memberNumber
        self.fullName = fullName
        self.fullNameAr = fullNameAr
        self.dateOfBirth = dateOfBirth
        self.gender = gender
        self.nationality = nationality
        self.emiratesID = emiratesID
        self.branchID = branchID
        self.primaryCoachID = primaryCoachID
        self.joinedAt = joinedAt
        self.currentBelt = currentBelt
        self.beltHistory = beltHistory
        self.weightKg = weightKg
        self.status = status
        self.avatarSeed = avatarSeed
        self.avatarURL = avatarURL
        self.passportNumber = passportNumber
        self.bloodType = bloodType
        self.federationLicenceNumber = federationLicenceNumber
        self.worldTaekwondoID = worldTaekwondoID
        self.parentUserIDs = parentUserIDs
        self.emergencyContacts = emergencyContacts
        self.school = school
        self.imageRightsConsent = imageRightsConsent
        self.imageRightsConsentDate = imageRightsConsentDate
        self.travelPermission = travelPermission
        self.travelPermissionDate = travelPermissionDate
        self.heightCm = heightCm
        self.weightHistory = weightHistory
        self.allergies = allergies
        self.medicalConditions = medicalConditions
        self.medications = medications
        self.fitToTrain = fitToTrain
        self.injuries = injuries
        self.weightClass = weightClass
        self.dominantLeg = dominantLeg
        self.dominantStance = dominantStance
        self.specialty = specialty
        self.yearsTraining = yearsTraining
        self.poomsaeSyllabus = poomsaeSyllabus
        self.kyorugiTier = kyorugiTier
        self.trainingGroupID = trainingGroupID
        self.poomsaeKnown = poomsaeKnown
        self.gradingReadiness = gradingReadiness.map { max(1, min(5, $0)) }
        self.nextGradingTargetDate = nextGradingTargetDate
        self.coachNotes = coachNotes
        self.documents = documents
        self.ranking = ranking
    }

    /// Whole months between when the current belt was awarded and now.
    public var monthsAtCurrentRank: Int {
        Calendar.current.dateComponents([.month], from: currentBelt.awardedAt, to: Date()).month ?? 0
    }

    public var age: Int {
        Calendar.current.dateComponents([.year], from: dateOfBirth, to: Date()).year ?? 0
    }

    public var ageGroup: AgeGroup { AgeGroup.from(age: age) }

    public var initials: String {
        let parts = fullName.split(separator: " ").prefix(2)
        return parts.map { String($0.prefix(1)) }.joined().uppercased()
    }

    /// Returns localization keys for fields that are recommended but missing.
    /// Empty result = profile is "complete enough". Used by the AthleteDetail
    /// warning banner to nudge the user without blocking creation.
    public var missingProfileFields: [String] {
        var missing: [String] = []
        if (emiratesID ?? "").isEmpty && (passportNumber ?? "").isEmpty {
            missing.append("athlete.missing.id_document")
        }
        if bloodType == nil { missing.append("athlete.missing.blood_type") }
        if (federationLicenceNumber ?? "").isEmpty { missing.append("athlete.missing.federation_licence") }
        if (avatarURL ?? "").isEmpty { missing.append("athlete.missing.photo") }
        if emergencyContacts.isEmpty { missing.append("athlete.missing.emergency_contact") }
        if (school ?? "").isEmpty { missing.append("athlete.missing.school") }
        if heightCm == nil { missing.append("athlete.missing.height") }
        if !imageRightsConsent { missing.append("athlete.missing.image_rights") }
        if !travelPermission { missing.append("athlete.missing.travel_permission") }
        return missing
    }

    /// 0...1, equally weighted across the recommended fields.
    public var profileCompleteness: Double {
        let totalChecked = 9.0
        let missingCount = Double(missingProfileFields.count)
        return max(0, min(1, (totalChecked - missingCount) / totalChecked))
    }
}

File: flutter_app/lib/core/models/athlete.dart
import 'athlete_extras.dart';
import 'athlete_profile.dart';
import 'belt.dart';
import 'coaching_development.dart';
import 'entity_id.dart';
import 'poomsae_assessment.dart' show PoomsaeForm;
import 'tournament.dart' show WeightCategory;

/// Port of Core/Models/Athlete.swift — full dossier.
///
/// Embeds the same per-athlete records the Swift struct does (emergency
/// contacts, injuries, weight history, coach notes, documents, ranking,
/// assistant-coach profile) as nested Codable fields. Decoding is tolerant of
/// missing keys (defaults applied) so Supabase rows and older seed data stay
/// forward-compatible — mirroring Swift's synthesised-Codable-with-defaults.

enum Gender {
  male,
  female;

  static Gender fromJson(String raw) =>
      Gender.values.firstWhere((g) => g.name == raw, orElse: () => Gender.male);
}

enum AgeGroup {
  cubs,
  kids,
  cadets,
  juniors,
  seniors,
  masters;

  String get labelKey => 'age.$name';

  static AgeGroup fromAge(int age) {
    if (age < 10) return AgeGroup.cubs;
    if (age <= 11) return AgeGroup.kids;
    if (age <= 14) return AgeGroup.cadets;
    if (age <= 17) return AgeGroup.juniors;
    if (age <= 39) return AgeGroup.seniors;
    return AgeGroup.masters;
  }

  static AgeGroup fromJson(String raw) => AgeGroup.values
      .firstWhere((e) => e.name == raw, orElse: () => AgeGroup.seniors);
}

enum AthleteStatus {
  competitionTeam,
  readyToGrade,
  watch,
  rest,
  active;

  String get labelKey => 'status.$name';

  static AthleteStatus fromJson(String raw) => AthleteStatus.values
      .firstWhere((s) => s.name == raw, orElse: () => AthleteStatus.active);
}

class Athlete {
  final EntityID id;
  final int memberNumber;
  final String fullName;
  final String fullNameAr;
  final DateTime dateOfBirth;
  final Gender gender;
  final String nationality;
  final String? emiratesID;
  final EntityID branchId;
  final EntityID? primaryCoachId;
  final DateTime joinedAt;
  final Belt currentBelt;
  final List<Belt> beltHistory;
  final double weightKg;
  final AthleteStatus status;
  final String avatarSeed;
  final String? avatarUrl;

  // === Identity ===
  final String? passportNumber;
  final BloodType? bloodType;
  final String? federationLicenceNumber;
  final String? worldTaekwondoID;

  // === Family / consent ===
  final List<EntityID> parentUserIDs;
  final List<EmergencyContact> emergencyContacts;
  final String? school;
  final bool imageRightsConsent;
  final DateTime? imageRightsConsentDate;
  final bool travelPermission;
  final DateTime? travelPermissionDate;

  // === Medical ===
  final double? heightCm;
  final List<WeightEntry> weightHistory;
  final List<String> allergies;
  final List<String> medicalConditions;
  final List<String> medications;
  final bool fitToTrain;
  final List<InjuryEntry> injuries;

  // === Technical ===
  final WeightCategory? weightClass;
  final DominantLeg? dominantLeg;
  final Stance? dominantStance;
  final Specialty? specialty;
  final int? yearsTraining;
  final String? poomsaeSyllabus;
  final EntityID? trainingGroupID;
  final Set<PoomsaeForm> poomsaeKnown;

  // === Grading ===
  final int? gradingReadiness; // clamped 1…5
  final DateTime? nextGradingTargetDate;

  // === Profile dossier ===
  final List<CoachNote> coachNotes;
  final List<AthleteDocument> documents;
  final AthleteRanking? ranking;

  // === Coaching pathway (Stage 1.15) ===
  final Set<ProgramRole> programRoles;
  final AssistantCoachProfile? assistantCoach;

  Athlete({
    required this.id,
    required this.memberNumber,
    required this.fullName,
    required this.fullNameAr,
    required this.dateOfBirth,
    this.gender = Gender.male,
    this.nationality = 'AE',
    this.emiratesID,
    required this.branchId,
    this.primaryCoachId,
    required this.joinedAt,
    required this.currentBelt,
    this.beltHistory = const [],
    required this.weightKg,
    required this.status,
    required this.avatarSeed,
    this.avatarUrl,
    this.passportNumber,
    this.bloodType,
    this.federationLicenceNumber,
    this.worldTaekwondoID,
    this.parentUserIDs = const [],
    this.emergencyContacts = const [],
    this.school,
    this.imageRightsConsent = false,
    this.imageRightsConsentDate,
    this.travelPermission = false,
    this.travelPermissionDate,
    this.heightCm,
    this.weightHistory = const [],
    this.allergies = const [],
    this.medicalConditions = const [],
    this.medications = const [],
    this.fitToTrain = true,
    this.injuries = const [],
    this.weightClass,
    this.dominantLeg,
    this.dominantStance,
    this.specialty,
    this.yearsTraining,
    this.poomsaeSyllabus,
    this.trainingGroupID,
    this.poomsaeKnown = const {},
    int? gradingReadiness,
    this.nextGradingTargetDate,
    this.coachNotes = const [],
    this.documents = const [],
    this.ranking,
    this.programRoles = const {ProgramRole.athlete},
    this.assistantCoach,
  }) : gradingReadiness =
            gradingReadiness == null ? null : (gradingReadiness.clamp(1, 5));

  /// Whole months between when the current belt was awarded and now.
  int get monthsAtCurrentRank {
    final now = DateTime.now();
    return (now.year - currentBelt.awardedAt.year) * 12 +
        (now.month - currentBelt.awardedAt.month);
  }

  int get age {
    final now = DateTime.now();
    var years = now.year - dateOfBirth.year;
    if (now.month < dateOfBirth.month ||
        (now.month == dateOfBirth.month && now.day < dateOfBirth.day)) {
      years -= 1;
    }
    return years;
  }

  AgeGroup get ageGroup => AgeGroup.fromAge(age);

  String get initials {
    final parts = fullName.split(' ').where((p) => p.isNotEmpty).take(2);
    return parts.map((p) => p.substring(0, 1)).join().toUpperCase();
  }

  /// Localization keys for recommended-but-missing fields. Empty = complete
  /// enough. Drives the AthleteDetail warning banner.
  List<String> get missingProfileFields {
    final missing = <String>[];
    if ((emiratesID ?? '').isEmpty && (passportNumber ?? '').isEmpty) {
      missing.add('athlete.missing.id_document');
    }
    if (bloodType == null) missing.add('athlete.missing.blood_type');
    if ((federationLicenceNumber ?? '').isEmpty) {
      missing.add('athlete.missing.federation_licence');
    }
    if ((avatarUrl ?? '').isEmpty) missing.add('athlete.missing.photo');
    if (emergencyContacts.isEmpty) {
      missing.add('athlete.missing.emergency_contact');
    }
    if ((school ?? '').isEmpty) missing.add('athlete.missing.school');
    if (heightCm == null) missing.add('athlete.missing.height');
    if (!imageRightsConsent) missing.add('athlete.missing.image_rights');
    if (!travelPermission) missing.add('athlete.missing.travel_permission');
    return missing;
  }

  /// 0…1, equally weighted across the recommended fields.
  double get profileCompleteness {
    const totalChecked = 9.0;
    final missingCount = missingProfileFields.length.toDouble();
    return ((totalChecked - missingCount) / totalChecked).clamp(0.0, 1.0);
  }

  factory Athlete.fromJson(Map<String, dynamic> json) => Athlete(
        id: json['id'] as String,
        memberNumber: json['memberNumber'] as int,
        fullName: json['fullName'] as String,
        fullNameAr: json['fullNameAr'] as String,
        dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
        gender: Gender.fromJson(json['gender'] as String? ?? 'male'),
        nationality: json['nationality'] as String? ?? 'AE',
        emiratesID: json['emiratesID'] as String?,
        branchId: json['branchID'] as String,
        primaryCoachId: json['primaryCoachID'] as String?,
        joinedAt: DateTime.parse(json['joinedAt'] as String),
        currentBelt: Belt.fromJson(json['currentBelt'] as Map<String, dynamic>),
        beltHistory: ((json['beltHistory'] as List?) ?? [])
            .map((e) => Belt.fromJson(e as Map<String, dynamic>))
            .toList(),
        weightKg: (json['weightKg'] as num).toDouble(),
        status: AthleteStatus.fromJson(json['status'] as String? ?? 'active'),
        avatarSeed: json['avatarSeed'] as String? ?? '',
        avatarUrl: json['avatarURL'] as String?,
        passportNumber: json['passportNumber'] as String?,
        bloodType: json['bloodType'] == null
            ? null
            : BloodType.fromJson(json['bloodType'] as String),
        federationLicenceNumber: json['federationLicenceNumber'] as String?,
        worldTaekwondoID: json['worldTaekwondoID'] as String?,
        parentUserIDs: ((json['parentUserIDs'] as List?) ?? [])
            .map((e) => e as String)
            .toList(),
        emergencyContacts: ((json['emergencyContacts'] as List?) ?? [])
            .map((e) => EmergencyContact.fromJson(e as Map<String, dynamic>))
            .toList(),
        school: json['school'] as String?,
        imageRightsConsent: json['imageRightsConsent'] as bool? ?? false,
        imageRightsConsentDate: json['imageRightsConsentDate'] == null
            ? null
            : DateTime.parse(json['imageRightsConsentDate'] as String),
        travelPermission: json['travelPermission'] as bool? ?? false,
        travelPermissionDate: json['travelPermissionDate'] == null
            ? null
            : DateTime.parse(json['travelPermissionDate'] as String),
        heightCm: (json['heightCm'] as num?)?.toDouble(),
        weightHistory: ((json['weightHistory'] as List?) ?? [])
            .map((e) => WeightEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
        allergies:
            ((json['allergies'] as List?) ?? []).map((e) => e as String).toList(),
        medicalConditions: ((json['medicalConditions'] as List?) ?? [])
            .map((e) => e as String)
            .toList(),
        medications: ((json['medications'] as List?) ?? [])
            .map((e) => e as String)
            .toList(),
        fitToTrain: json['fitToTrain'] as bool? ?? true,
        injuries: ((json['injuries'] as List?) ?? [])
            .map((e) => InjuryEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
        weightClass: json['weightClass'] == null
            ? null
            : WeightCategory.fromJson(json['weightClass'] as String),
        dominantLeg: json['dominantLeg'] == null
            ? null
            : DominantLeg.fromJson(json['dominantLeg'] as String),
        dominantStance: json['dominantStance'] == null
            ? null
            : Stance.fromJson(json['dominantStance'] as String),
        specialty: json['specialty'] == null
            ? null
            : Specialty.fromJson(json['specialty'] as String),
        yearsTraining: json['yearsTraining'] as int?,
        poomsaeSyllabus: json['poomsaeSyllabus'] as String?,
        trainingGroupID: json['trainingGroupID'] as String?,
        poomsaeKnown: ((json['poomsaeKnown'] as List?) ?? [])
            .map((e) => PoomsaeForm.fromJson(e as String))
            .toSet(),
        gradingReadiness: json['gradingReadiness'] as int?,
        nextGradingTargetDate: json['nextGradingTargetDate'] == null
            ? null
            : DateTime.parse(json['nextGradingTargetDate'] as String),
        coachNotes: ((json['coachNotes'] as List?) ?? [])
            .map((e) => CoachNote.fromJson(e as Map<String, dynamic>))
            .toList(),
        documents: ((json['documents'] as List?) ?? [])
            .map((e) => AthleteDocument.fromJson(e as Map<String, dynamic>))
            .toList(),
        ranking: json['ranking'] == null
            ? null
            : AthleteRanking.fromJson(json['ranking'] as Map<String, dynamic>),
        programRoles: json['programRoles'] == null
            ? const {ProgramRole.athlete}
            : (json['programRoles'] as List)
                .map((e) => ProgramRole.fromJson(e as String))
                .toSet(),
        assistantCoach: json['assistantCoach'] == null
            ? null
            : AssistantCoachProfile.fromJson(
                json['assistantCoach'] as Map<String, dynamic>),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'memberNumber': memberNumber,
        'fullName': fullName,
        'fullNameAr': fullNameAr,
        'dateOfBirth': dateOfBirth.toIso8601String(),
        'gender': gender.name,
        'nationality': nationality,
        'emiratesID': emiratesID,
        'branchID': branchId,
        'primaryCoachID': primaryCoachId,
        'joinedAt': joinedAt.toIso8601String(),
        'currentBelt': currentBelt.toJson(),
        'beltHistory': beltHistory.map((b) => b.toJson()).toList(),
        'weightKg': weightKg,
        'status': status.name,
        'avatarSeed': avatarSeed,
        'avatarURL': avatarUrl,
        'passportNumber': passportNumber,
        'bloodType': bloodType?.rawValue,
        'federationLicenceNumber': federationLicenceNumber,
        'worldTaekwondoID': worldTaekwondoID,
        'parentUserIDs': parentUserIDs,
        'emergencyContacts': emergencyContacts.map((e) => e.toJson()).toList(),
        'school': school,
        'imageRightsConsent': imageRightsConsent,
        'imageRightsConsentDate': imageRightsConsentDate?.toIso8601String(),
        'travelPermission': travelPermission,
        'travelPermissionDate': travelPermissionDate?.toIso8601String(),
        'heightCm': heightCm,
        'weightHistory': weightHistory.map((e) => e.toJson()).toList(),
        'allergies': allergies,
        'medicalConditions': medicalConditions,
        'medications': medications,
        'fitToTrain': fitToTrain,
        'injuries': injuries.map((e) => e.toJson()).toList(),
        'weightClass': weightClass?.name,
        'dominantLeg': dominantLeg?.name,
        'dominantStance': dominantStance?.name,
        'specialty': specialty?.name,
        'yearsTraining': yearsTraining,
        'poomsaeSyllabus': poomsaeSyllabus,
        'trainingGroupID': trainingGroupID,
        'poomsaeKnown': poomsaeKnown.map((e) => e.name).toList(),
        'gradingReadiness': gradingReadiness,
        'nextGradingTargetDate': nextGradingTargetDate?.toIso8601String(),
        'coachNotes': coachNotes.map((e) => e.toJson()).toList(),
        'documents': documents.map((e) => e.toJson()).toList(),
        'ranking': ranking?.toJson(),
        'programRoles': programRoles.map((e) => e.name).toList(),
        'assistantCoach': assistantCoach?.toJson(),
      };
}
